core/services/job/orm.go

```diff
@@ -845,42 +845,98 @@ func (o *orm) PipelineRunsByJobsIDs(ids []int32) (runs []pipeline.Run, err error
 }
 
+func (o *orm) loadPipelineRunIDs(jobID *int32, offset, limit int, tx pg.Queryer) (ids []int64, err error) {
+	var filter string
+
+	if jobID != nil {
+		filter = "JOIN jobs USING(pipeline_spec_id) WHERE jobs.id = $1 AND "
+	} else {
+		filter = "WHERE "
+	}
+	stmt := fmt.Sprintf(`SELECT p.id FROM pipeline_runs AS p %s p.id >= $4 AND p.id <= $5
+			ORDER BY p.id DESC OFFSET $2 LIMIT $3`, filter)
+	var res sql.NullInt64
+
+	if err = tx.Get(&res, "SELECT MAX(id) FROM pipeline_runs"); err != nil {
+		err = errors.Wrap(err, "error while loading runs")
+		return
+	} else if !res.Valid {
+		// MAX() will return NULL if there are no rows in table.  This is not an error
+		return
+	}
+
+	maxID := res.Int64
+
+	// Only search the most recent n jobs (whether deleted or not), starting with n = 1000 and
+	//  doubling only if we still need more.  Without this, large tables can result in the UI
+	//  becoming unusably slow, continuously flashing, or timing out.  The ORDER BY in
+	//  this query requires a sort of all runs matching jobID, so we restrict it to the
+	//  range minID <-> maxID.
+
+	for n := int64(1000); maxID > 0 && len(ids) < limit; n *= 2 {
+		minID := maxID - n
+		if err = tx.Select(&ids, stmt, jobID, offset, limit-len(ids), minID, maxID); err != nil {
+			err = errors.Wrap(err, "error loading runs")
+			return
+		}
+		if offset > 0 {
+			if len(ids) > 0 {
+				// If we're already receiving rows back, then we no longer need an offset
+				offset = 0
+			} else {
+				var skipped int
+				// If no rows were returned, we need to know whether there were any ids skipped
+				//  in this batch due to the offset, and reduce it for the next batch
+				err = tx.Select(&skipped,
+					fmt.Sprintf(
+						`SELECT COUNT(p.id) FROM pipeline_runs AS p %s p.id >= $2 AND p.id <= $3`, filter,
+					),
+				)
+				if err != nil {
+					err = errors.Wrap(err, "error loading from pipeline_runs")
+					return
+				}
+				offset -= skipped
+				if offset < 0 { // sanity assertion, if this ever happened it would probably mean db corruption or pg bug
+					err = errors.Wrap(err, "internal db error while reading pipeline_runs")
+					return
+				}
+				o.lggr.Debugw("loadPipelineRunIDs empty batch", "minId", minID, "maxID", maxID, "n", n, "len(ids)", len(ids), "limit", limit, "offset", offset, "skipped", skipped)
+
+			}
+		}
+		maxID = minID - 1
+	}
+	return
+}
+
 // FindPipelineRunIDsByJobID fetches the ids of pipeline runs for a job.
 func (o *orm) FindPipelineRunIDsByJobID(jobID int32, offset, limit int) (ids []int64, err error) {
 	err = o.q.Transaction(func(tx pg.Queryer) error {
-		stmt := `
-SELECT pipeline_runs.id
-FROM pipeline_runs
-WHERE pipeline_runs.pipeline_spec_id = (SELECT jobs.pipeline_spec_id FROM JOBS WHERE jobs.id = $1)
-ORDER BY pipeline_runs.created_at DESC, pipeline_runs.id DESC
-OFFSET $2
-LIMIT $3
-`
-		if err = tx.Select(&ids, stmt, jobID, offset, limit); err != nil {
-			return errors.Wrap(err, "error loading runs")
-		}
-
```

Comment:
This was the slow part for jobs with large numbers of pipeline runs? I am just surprised there was no way to use an index to make this query performant.

Reply:
Yes, this was the bottleneck, and became even more of a bottleneck after getting it to use the index (by switching to BTREE from BRIN).... although using the index did help some.

In the case where there is no job id specified, the index was enough to make it fast.  The real problem came when there was a specific job id passed.

If there are 10 jobs, each with 500,000 runs, then consider how much work has to be done for the case where you want to fetch 10 runs of job_id=3 with offset=450,000 limit=450,010.  It has to scan through all 5-million rows, finding and sorting the first 450,000 it finds which match job_id=3.  The BRIN can't help at all with this, since it just stores the lowest and highest timestamp for each 1MB block of data on the disk.  Inside each 1MB block, nothing is sorted or indexable in any particular order.  The BTREE index fixed the problem when there is no job id specified, because then everything is sorted in a btree in the order you're accessing it.  But in the case of multiple job id's where you only care about one, it's still got a lot of work to do.  I was hoping that would be enough, but after testing it I found that it was still pretty slow (much slower than the COUNT query, for instance).

Reply:
Thanks for this explanation